### PR TITLE
add copy_from parameter for set ingest processor

### PR DESCRIPTION
**Pull request by gaobinlong**
Relates to #55682 and #51046, and refractor the code of #56985. 

The main point of this PR is adding a `copy_from` parameter for set ingest processor to support `copy` the value of one field to another. We can copy `boolean`, `number`, `array`, `object` and other data types by the parameter `copy_from` instead of using `template snippet` which only produces `string` value. 

Why adding a new parameter to implement the `copy` function but not modifying the `template snippet` so that it can produce `object` or `array` value ? I think we can make it compatible with older versions, users can still use `template snippet` to copy the produced `string` value to other field, but also can use `copy_from` to copy `object` or `array` value from one field to another.

The main changes are:
1. Add `copy_from` parameter for set ingest processor.
2. Add unit test and yml test for the change.
3. Update the set processor's document.

### CODE REVIEW

**Comment by danhermann** [comment]
> Why adding a new parameter to implement the `copy` function but not modifying the `template snippet` so that it can produce `object` or `array` value ? I think we can make it compatible with older versions, users can still use `template snippet` to copy the produced `string` value to other field, but also can use `copy_from` to copy `object` or `array` value from one field to another.

Would it be possible to modify the `template snippet` code to produce `object` and `array` values? Our team discussed this and couldn't think of any good reason for producing `string` values. We would release this feature in the next major version of ES so maintaining compatibility with older versions is not required.

**Comment by gaobinlong** [comment]
@danhermann, the usage of `template snippet` is flexible, for example, `"{{field_a}} {{field_b}}"` is used to concatenate the value of `field_a` and `field_b`, but if `field_a` and `field_b` are not `string`, what's the result after executing the `template snippet` if `template snippet` can produce `object` or `array` value?  I'm not clear about that. Another example is that `{{.}}` is valid currently which will output the whole document. So maybe specifying  a field by the parameter `copy_from` can simply copy the value from one filed to another without handing some complex situation. In addition, maybe it's not easy to modify the `template snippet` code to produce object and array values, from the source code: [MustacheScriptEngine](https://github.com/elastic/elasticsearch/blob/213c627c35893cde88416b12e2e70f453d4bdea2/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java#L124), I found that the `MustacheScriptEngine` only produces `string` value. Expect your opinion on this.

**Comment by danhermann** [comment]
> @danhermann, the usage of `template snippet` is flexible, for example, `"{{field_a}} {{field_b}}"` is used to concatenate the value of `field_a` and `field_b`, but if `field_a` and `field_b` are not `string`, what's the result after executing the `template snippet` if `template snippet` can produce `object` or `array` value? I'm not clear about that. Another example is that `{{.}}` is valid currently which will output the whole document. So maybe specifying a field by the parameter `copy_from` can simply copy the value from one filed to another without handing some complex situation. In addition, maybe it's not easy to modify the `template snippet` code to produce object and array values, from the source code: [MustacheScriptEngine](https://github.com/elastic/elasticsearch/blob/213c627c35893cde88416b12e2e70f453d4bdea2/modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngine.java#L124), I found that the `MustacheScriptEngine` only produces `string` value. Expect your opinion on this.

Thanks, @gaobinlong. Your explanation makes sense and we'd like to move forward with the `copy_from` parameter as you've implemented it. I'll review this shortly and we'll work on getting it merged.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine ok to test

File: docs/reference/ingest/processors/set.asciidoc [hunk 1]
```diff
@@ -13,7 +13,8 @@ its value will be replaced with the provided one.
 |======
 | Name       | Required  | Default  | Description
 | `field` | yes       | -        | The field to insert, upsert, or update. Supports <<accessing-template-fields,template snippets>>.
-| `value` | yes       | -        | The value to be set for the field. Supports <<accessing-template-fields,template snippets>>.
+| `value` | no       | -        | The value to be set for the field. Supports <<accessing-template-fields,template snippets>>.
```
Comment (danhermann) [inline docs/reference/ingest/processors/set.asciidoc:16]:
```suggestion
| `value` | yes*     | -        | The value to be set for the field. Supports <<accessing-template-fields,template snippets>>. May specify only one of `value` or `copy_from`.
```

File: docs/reference/ingest/processors/set.asciidoc [hunk 2]
```diff
@@ -87,3 +88,54 @@ Result:
 }
 --------------------------------------------------
 // TESTRESPONSE[s/2019-03-11T21:54:37.909224Z/$body.docs.0.doc._ingest.timestamp/]
+As `template snippets` only produces string value, we can copy an array from one field to another by using `copy_from`:
```
Comment (danhermann) [inline docs/reference/ingest/processors/set.asciidoc:91]:
```suggestion
The contents of a field including complex values such as arrays and objects can be copied to another field using `copy_from`:
```

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/SetProcessor.java [hunk 3]
```diff
@@ -96,16 +109,29 @@ public Factory(ScriptService scriptService) {
...
-            Object value = ConfigurationUtils.readObject(TYPE, processorTag, config, "value");
+            String copyFrom = ConfigurationUtils.readOptionalStringProperty(TYPE, processorTag, config, "copy_from");
+            ValueSource valueSource = null;
+            if (copyFrom == null) {
+                Object value = ConfigurationUtils.readObject(TYPE, processorTag, config, "value");
+                valueSource = ValueSource.wrap(value, scriptService);
+            } else {
+                Object value = config.remove("value");
+                if (value != null) {
+                    throw newConfigurationException(TYPE, processorTag, "copy_from", "cannot set `copy_from` while also setting `value`");
```
Comment (danhermann) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/SetProcessor.java:120]:
```suggestion
                    throw newConfigurationException(TYPE, processorTag, "copy_from", "cannot set both `copy_from` and `value` in the same processor");
```

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/SetProcessorFactoryTests.java [hunk 4]
```diff
@@ -112,4 +112,21 @@ public void testInvalidMustacheTemplate() throws Exception {
...
+        assertThat(setProcessor.getTag(), equalTo(processorTag));
+        assertThat(setProcessor.getField().newInstance(Collections.emptyMap()).execute(), equalTo("field1"));
+        assertThat(setProcessor.getCopyFrom(), equalTo("field2"));
+
+        config.put("field", "field1");
+        config.put("value", "value1");
+        config.put("copy_from", "field2");
+        ElasticsearchException exception = expectThrows(ElasticsearchException.class,
+            () -> factory.create(null, processorTag, null, config));
+        assertThat(exception.getMessage(), equalTo("[copy_from] cannot set `copy_from` while also setting `value`"));
```
Comment (danhermann) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/SetProcessorFactoryTests.java:130]: Can you make a second test method for this test case to make it clear that setting both `value` and `copy_from` is an error that should be caught by the factory at creation time?

**Review by danhermann — CHANGES_REQUESTED** [review]
@gaobinlong, from my testing, this looks good and works well. I left some minor wording changes below as well as a request to split up one of the unit tests. I think we can get it merged after that.

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/SetProcessorFactoryTests.java (hunk 4 above)
Comment (gaobinlong) [reply modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/SetProcessorFactoryTests.java]: OK, I will do that.

**Comment by gaobinlong** [comment]
@elasticmachine test this please.

**Comment by danhermann** [comment]
Thanks, @gaobinlong. This looks good. The last failing test is due to the fact that line 120 in `SetProcessor.java` is now longer than the limit of 140 characters. If you can split the line to keep it under the length limit, I believe that all tests will pass.

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
Thanks again for this contribution, @gaobinlong! I'll get it merged and backported.

**Pull request merged**